pkg/util/globalerror/grpc.go

```diff
@@ -0,0 +1,161 @@
+	"google.golang.org/grpc/codes"
+	grpcstatus "google.golang.org/grpc/status"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+// WrapGrpcContextError wraps a gRPC error with a custom error mapping gRPC to standard golang context errors.
+func WrapGrpcContextError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	// Get the gRPC status from the error.
+	type grpcErrorStatus interface{ GRPCStatus() *grpcstatus.Status }
+	var grpcError grpcErrorStatus
+	if !errors.As(err, &grpcError) {
+		return err
+	}
+
+	switch status := grpcError.GRPCStatus(); {
+	case status.Code() == codes.Canceled:
+		return &grpcContextError{grpcErr: err, grpcStatus: status, stdErr: context.Canceled}
+	case status.Code() == codes.DeadlineExceeded:
+		return &grpcContextError{grpcErr: err, grpcStatus: status, stdErr: context.DeadlineExceeded}
+	default:
+		return err
+	}
+}
+
+// grpcContextError is a custom error used to wrap gRPC errors which maps to golang standard context errors.
+type grpcContextError struct {
+	// grpcErr is the gRPC error wrapped by grpcContextError.
+	grpcErr error
+
+	// grpcStatus is the gRPC status associated with the gRPC error. It's guaranteed to be non-nil.
+	grpcStatus *grpcstatus.Status
+
+	// stdErr is the equivalent golang standard context error.
+	stdErr error
+}
+
+func (e *grpcContextError) Error() string {
+	return e.grpcErr.Error()
+}
+
+func (e *grpcContextError) Unwrap() error {
+	return e.grpcErr
+}
+
+func (e *grpcContextError) Is(err error) bool {
+	return errors.Is(e.stdErr, err) || errors.Is(e.grpcErr, err)
+}
+
+func (e *grpcContextError) As(target any) bool {
+	if errors.As(e.stdErr, target) {
+		return true
+	}
+
+	return errors.As(e.grpcErr, target)
+}
+
+func (e *grpcContextError) GRPCStatus() *grpcstatus.Status {
+	return e.grpcStatus
+}
+
+// ErrorWithStatus is used for wrapping errors returned by ingester.
+// Errors returned by ingester should be gRPC errors, but the errors
+// produced by both gogo/status and grpc/status packages do not keep
+// the semantics of the underlying error, which is sometimes needed.
+// For example, the logging middleware needs to know whether an error
+// should be logged, sampled or ignored. Errors of type ErrorWithStatus
+// are valid gRPC errors that could be parsed by both gogo/status
+// and grpc/status packages, but which preserve the original error
+// semantics.
+type ErrorWithStatus struct {
+	UnderlyingErr error
+	Status        *status.Status
+}
+
+func NewErrorWithGRPCStatus(originalErr error, code codes.Code, details *mimirpb.ErrorDetails) ErrorWithStatus {
+	stat := status.New(code, originalErr.Error())
+	if details != nil {
+		if statWithDetails, err := stat.WithDetails(details); err == nil {
+			return ErrorWithStatus{
+				UnderlyingErr: originalErr,
+				Status:        statWithDetails,
+			}
+		}
+	}
+	return ErrorWithStatus{
+		UnderlyingErr: originalErr,
+		Status:        stat,
+	}
+}
+
+func (e ErrorWithStatus) Error() string {
+	return e.Status.Message()
```

Comment:
Look at the code changes in `distributor. toGRPCError()`. I've the feeling the `Error()` we'll return from this function will be different. Previously it was gRPC `statusError.Error()` but now it's the status message (`statusError.Error()` decorates the status message). I'm not sure whether this is a problem or not, but if my analysis is correct I think we're masquerading a change in a refactoring PR.

Reply:
> I've the feeling the `Error()` we'll return from this function will be different

you're right. We're now [returning `globalerror.ErrorWithStatus`](https://github.com/grafana/mimir/blob/069ef53f931682c4fc5e289faa4373d85c399ffc/pkg/util/globalerror/grpc.go#L93) instead of [`status.*statusError`](https://github.com/grafana/mimir/blob/7b2283ff8f296fd231bcefae64386524842e5ed9/vendor/github.com/gogo/status/status.go#L104).

But also the actual error **strings**  will be something like `connection refused` - because `globalerror.ErrorWithStatus` returns purely the message https://github.com/grafana/mimir/blob/069ef53f931682c4fc5e289faa4373d85c399ffc/pkg/util/globalerror/grpc.go#L105-L107. Whereas previously they were `rpc error: code = Unavailable desc = connection refused` because that's what [`status.*statusError` constructs](https://github.com/grafana/mimir/blob/7b2283ff8f296fd231bcefae64386524842e5ed9/vendor/github.com/gogo/status/status.go#L49)

i still think it's only a cosmetic change, but I can't be certain, so I'll revert that change and leave a comment if anyone feels adventurous in the future

Reply:
restored the previous code and added a comment in 7713def310ee99af55c2d26a0bb9b191c4d05a19

Reply:
> i still think it's only a cosmetic change, but I can't be certain, so I'll revert that change and leave a comment if anyone feels adventurous in the future

Hi @dimitarvdimitrov, 
Since I was not involved in reviewing PR, I will reply you now. It is **not** a cosmetic change.

### TL;DR
Errors of type `globalerror.ErrorWithStatus` are gRPC errors that could be recognized by both grpc and gogo libraries, that preserve the semantics of their underlying error. Regular gRPC errors created by both grpc and gogo libraries do not have that features. Errors produced by ingesters should preserve the underlying semantics, while errors produced by distributors don't have to preserve it (at least not in this moment).

### Details 
Taken from the [current](https://github.com/grafana/mimir/blob/1f2028dbfdaf8e7ba15173966a0dba08314a7c35/pkg/distributor/errors.go#L260-L263) `distributor/errors.go`:
```
// errorWithStatus is slightly different from globalerror.NewErrorWithGRPCStatus. It returns status.Err() instead of an error with the status message.
// The actual difference is between "rpc error: code = XYZ desc = message" and just "message".
// At the time of writing this should be purely a cosmetic difference, but it's hard to verify.
// Because of this difference, the function is used instead of globalerror.NewErrorWithGRPCStatus.
```

[This](https://github.com/grafana/mimir/pull/7790/files#diff-202b322034387f4af9d245d4087160d23441deb864332b433cd004920ae6ebe4R256-R272) is the change you did in `distributor.toGRPCError()`. You basically cut one piece of the code and created a function out of it. This function simply creates _a gRPC error with details_ by using gogo's `status` package: error message corresponds to the original errors's `Error()`, and the detail is the error cause of type `mimirpb.ErrorCause`. This way **no information about original error's semantics** is preserved. For example, if a grpc logging middleware receives this error, it cannot understand whether it was a non-loggable error, or an error that should be sampled etc. Just its error message and its details (used for a communication between different Mimir components) would be sent.

I recommend that we either revert [this change](https://github.com/grafana/mimir/pull/7790/files#diff-202b322034387f4af9d245d4087160d23441deb864332b433cd004920ae6ebe4R256-R272) or that we remove the comment which introduces confusion (it looks like we are not so sure why we do it this way).

On the other hand, we have the [current](https://github.com/grafana/mimir/blob/2869d0a18cb3cd9ac92670e0cb928fc2a259e749/pkg/util/globalerror/grpc.go#L75-L83) `globalerror.ErrorWithStatus`:
```
// ErrorWithStatus is used for wrapping errors returned by ingester.
// Errors returned by ingester should be gRPC errors, but the errors
// produced by both gogo/status and grpc/status packages do not keep
// the semantics of the underlying error, which is sometimes needed.
// For example, the logging middleware needs to know whether an error
// should be logged, sampled or ignored. Errors of type ErrorWithStatus
// are valid gRPC errors that could be parsed by both gogo/status
// and grpc/status packages, but which preserve the original error
// semantics.
```

I put this comment when I implemented `ErrorWithStatus` within the `ingester` package in order to explain its main purpose, and why it was different from regular gRPC errors created by the two libraries. I think that the explanation (already present in the code) is clear. But if it is not we can rewrite it better.